File: Server/src/main/content/global/skill/production/crafting/data/GemData.java
package content.global.skill.production.crafting.data;

import core.game.node.item.Item;
import core.game.world.update.flag.context.Animation;

public enum GemData {
    SAPPHIRE(new Item(1623), new Item(1607), 20, new Animation(888), 50),
    EMERALD(new Item(1621), new Item(1605), 27, new Animation(889), 67),
    RUBY(new Item(1619), new Item(1603), 34, new Animation(887), 85),
    DIAMOND(new Item(1617), new Item(1601), 43, new Animation(886), 107.5),
    DRAGONSTONE(new Item(1631), new Item(1615), 55, new Animation(885), 137.5),
    ONYX(new Item(6571), new Item(6573), 67, new Animation(2717), 168),
    OPAL(new Item(1625), new Item(1609), 1, new Animation(890), 10),
    JADE(new Item(1627), new Item(1611), 13, new Animation(891), 20),
    RED_TOPAZ(new Item(1629), new Item(1613), 16, new Animation(892), 25);

    private final Item uncut;
    private final Item gem;
    private final int level;
    private final Animation animation;
    private final double exp;

    GemData(Item uncut, Item gem, int level, Animation animation, double exp) {
        this.uncut = uncut;
        this.gem = gem;
        this.level = level;
        this.animation = animation;
        this.exp = exp;
    }

    public static GemData forId(final Item item) {
        for (GemData gem : GemData.values()) {
            if (gem.getUncut().getId() == item.getId()) {
                return gem;
            }
        }
        return null;
    }

    public Item getUncut() {
        return uncut;
    }
    public Item getGem() {
        return gem;
    }
    public int getLevel() {
        return level;
    }
    public Animation getAnimation() {
        return animation;
    }
    public double getExp() {
        return exp;
    }
}

File: Server/src/main/content/global/skill/production/crafting/data/GemData.kt
package content.global.skill.production.crafting.data

import core.api.consts.Animations
import core.api.consts.Items
import core.game.node.item.Item
import core.game.world.update.flag.context.Animation

/**
 * Gem data
 *
 * @property uncut
 * @property gem
 * @property level
 * @property animation
 * @property exp
 * @constructor Gem data
 */
enum class GemData(val uncut: Item, val gem: Item, val level: Int, val animation: Animation, val exp: Double) {
    /**
     * Sapphire
     *
     * @constructor Sapphire
     */
    SAPPHIRE(
        uncut = Item(Items.UNCUT_SAPPHIRE_1623),
        gem = Item(Items.SAPPHIRE_1607),
        level = 20,
        animation = Animation(Animations.CUT_SAPPHIRE_888),
        exp = 50.0
    ),

    /**
     * Emerald
     *
     * @constructor Emerald
     */
    EMERALD(
        uncut = Item(Items.UNCUT_EMERALD_1621),
        gem = Item(Items.EMERALD_1605),
        level = 27,
        animation = Animation(Animations.CUT_EMERALD_889),
        exp = 67.0
    ),

    /**
     * Ruby
     *
     * @constructor Ruby
     */
    RUBY(
        uncut = Item(Items.UNCUT_RUBY_1619),
        gem = Item(Items.RUBY_1603),
        level = 34,
        animation = Animation(Animations.CUT_RUBY_887),
        exp = 85.0
    ),

    /**
     * Diamond
     *
     * @constructor Diamond
     */
    DIAMOND(
        uncut = Item(Items.UNCUT_DIAMOND_1617),
        gem = Item(Items.DIAMOND_1601),
        level = 43,
        animation = Animation(Animations.CUT_DIAMOND_886),
        exp = 107.5
    ),

    /**
     * Dragonstone
     *
     * @constructor Dragonstone
     */
    DRAGONSTONE(
        uncut = Item(Items.UNCUT_DRAGONSTONE_1631),
        gem = Item(Items.DRAGONSTONE_1615),
        level = 55,
        animation = Animation(Animations.CUT_DRAGONSTONE_885),
        exp = 137.5
    ),

    /**
     * Onyx
     *
     * @constructor Onyx
     */
    ONYX(
        uncut = Item(Items.UNCUT_ONYX_6571),
        gem = Item(Items.ONYX_6573),
        level = 67,
        animation = Animation(Animations.CHISEL_ONYX_2717),
        exp = 168.0
    ),

    /**
     * Opal
     *
     * @constructor Opal
     */
    OPAL(
        uncut = Item(Items.UNCUT_OPAL_1625),
        gem = Item(Items.OPAL_1609),
        level = 1,
        animation = Animation(Animations.CUT_OPAL_890),
        exp = 10.0
    ),

    /**
     * Jade
     *
     * @constructor Jade
     */
    JADE(
        uncut = Item(Items.UNCUT_JADE_1627),
        gem = Item(Items.JADE_1611),
        level = 13,
        animation = Animation(Animations.CUT_JADE_891),
        exp = 20.0
    ),

    /**
     * Red Topaz
     *
     * @constructor Red Topaz
     */
    RED_TOPAZ(
        uncut = Item(Items.UNCUT_RED_TOPAZ_1629),
        gem = Item(Items.RED_TOPAZ_1613),
        level = 16,
        animation = Animation(Animations.CUT_TOPAZ_892),
        exp = 25.0
    );

    companion object {
        fun forId(item: Item): GemData? {
            for (gem in values()) {
                if (gem.uncut.id == item.id) {
                    return gem
                }
            }
            return null
        }
    }
}
